### DISCUSSION TITLE
Documentation about using 'attr_dict=' with '.add_node()'?

### DISCUSSION THREAD

**Discussion by MikeB2019x** [networkx/networkx#6668] [General]
Was showing a colleague around networkx and they asked about my use of `attr_dict` as an argument for the `add_node()` method.  For example this works in `v2.8.7`:
```
G_.add_node('n75095', attr_dict={'TYPE': ':SalesOrder', 'date': '2022-05-13'})
```
They pointed out that they didn't see that in the docs and I honestly can't remember where I got the approach from.  A search in the docs and code shows references to using `attr_dict` but in the context of it not being allowed for `.add_edges()` method since v2.0.  A look at the code from v2.8.7 and the `main` branch appear the same.  I find references to it's use on Stack.  Are there docs or is it a hack or have I just embarrassed myself by asking?

**Comment by rossbar** [comment]
There is no `attr_dict` keyword argument or anything - node/edge attributes can be added by keyword directly, e.g.

```python
>>> G = nx.Graph()
>>> G.add_node("n75095", TYPE="SalesOrder", date="2022-05-13")
>>> G.nodes(data=True)
NodeDataView({'n75095': {'TYPE': 'SalesOrder', 'date': '2022-05-13'}})
```

Note that this is equivalent to unpacking the dict from your original example, e.g. `G.add_node("n75095", **attr_dict)`.

Of course, it's perfectly valid to add a dictionary as a node attribute (named however you like), you'll just end up with another layer of dict-nesting when accessing your attributes!

**Reply by MikeB2019x** [reply]
Thanks!  I've been using that blindly for years.  It _is_ super convenient but yeah another layer of dict that I just accepted as how life was :-D Though it would be handy to be able to just drop in a dict.

**Comment by dschult** [comment]
Yes -- you can drop in a dict by using the `**` python feature:
```G.add_node('n75095', **{'TYPE': ':SalesOrder', 'date': '2022-05-13'})```
And if you have the dict and it is named `attr_dict` then it is:
```G.add_node('n75095', **attr_dict)```
But it may be smoother when you don't have the dict already to use:
```G.add_node('n75095', TYPE='SalesOrder', date= '2022-05-13')```

Note that the way you were using means you access it with:  `G.nodes['n75095']['attr_dict']['TYPE']`
With any of the ways in this comment you would use:  `G.nodes['n75095']['TYPE']`
